README.md
# kano-settings

[![Coverage](http://dev.kano.me/public/status-badges/kano-settings-coverage.svg)](https://jenkins.kano.me/job/Core%20Kit/job/kano-settings/job/master/cobertura)

Kano Settings is like a Swiss Army knife: crammed with lots of useful applications, all conveniently packaged together for ease of use and topped off with a lovely UI. It’s a great tool, especially for beginners, to get started quickly without having to tinker with different config files. Set up your wifi, keyboard, mouse, audio, display, account, wallpaper, and overclock your Raspberry Pi, all with one handy utility.

## Settings


### Keyboard

This setting configures the keyboard given a country and a variant.

### Mouse

Configuration of the mouse speed: low, normal and high.

### Audio

Audio output can be set to Analog (speaker or headphones) or HDMI (tv).

### Display

The resolution of the display can be configured from this setting. The list of modes offered depend directly on the screen/monitor used.

If your screen requires special settings or lacks EDID information and you need to manually set them,
create a file in the boot partition as well with the name `no-config-writes`. This will tell Kano OS not to try to optimize
your modes or resolution.

### WiFi

Tool to configure your internet connection: ethernet, wifi or smartphone USB tethering.

### Overclocking

Get the most of your RaspberryPi. Choose from None, Modest, Medium or High.

### Account

Change your password, remove your account or add a new one.

### Wallpaper

Select a wallpaper for your desktop.


debian/changelog
kano-settings (4.3.3-0) unstable; urgency=low

  * Moved the DNS Sentry server to a system systemd service
    and refactored lib

 -- Team Kano <dev@kano.me>  Mon, 16 Sep 2019 11:15:00 +0100

kano-settings (4.3.2-0) unstable; urgency=low

  * Added parental settings to kano-settings-cli

 -- Team Kano <dev@kano.me>  Fri, 30 Aug 2019 13:15:00 +0100

kano-settings (4.2.1-0) unstable; urgency=low

  * Fixed race condition at boot preventing Wi-Fi from reconnecting to existing network

 -- Team Kano <dev@kano.me>  Wed, 16 Jan 2019 13:34:00 +0100

kano-settings (4.2.0-0) unstable; urgency=low

  * Fixed an issue in Parental Controls when going from Ultimate
    to High and then selecting to configure the security level
  * Fixed corrupt config.txt recovery errors when file is missing
  * Fixed tbk caused by not closing config transaction during
    kano-settings-onboot
  * Enable styling settings on Dashboard
  * Migrated Safe Mode functionality to its own project
  * Added config_copy_to() and config_copy_from() transaction functions

 -- Team Kano <dev@kano.me>  Tue, 20 Nov 2018 12:58:00 +0000

kano-settings (4.1.0-0) unstable; urgency=low

  * Added a simplistic system integrity script to verify debian packages.
  * Removed default RPi2 CPU frequency settings from config.txt
  * Mechanism to ensure SK display orientation consistency if EDID is lost
  * Fix inconsistent Settings > Display > Resolution: Auto behaviour for
    displays with corrupted EDID
  * Switch to use the "dtparam" value to check for config.txt corruption
  * Fix corrupt config.txt logging message

 -- Team Kano <dev@kano.me>  Thu, 27 Sep 2018 12:11:00 +0100

kano-settings (4.0.6-0) unstable; urgency=low

  * Changes for Kano OS 4.0.0:
  * Fixed an issue with automatic resolution setting for screens with aspect
    ratios that are not 4:3, 16:9, 16:10
  * Fix support for Gtk packages included with Stretch
  * Fixed Safe Mode configuration sometimes not getting applied correctly
  * Changed to require immediate reboot when adding an account.
  * Avoid traceback if bluetooth service failed to start.

 -- Team Kano <dev@kano.me>  Fri, 22 Jun 2018 10:17:00 +0100

kano-settings (4.0.5-0) unstable; urgency=low

  * Changes for Kano OS 3.15.0:
  * Fixed long delay when going into the Notifications menu
  * Disabled hdmi_drive config option when going into Safe Mode
  * Add splash animation for Settings app
  * Fixed the Wi-Fi refresh screen on connection failure
  * Create test coverage report
  * Run tests on push
  * Fixed applying setting when clicking on Flip Screen option

 -- Team Kano <dev@kano.me>  Wed, 24 Jan 2018 14:15:00 +0100

kano-settings (4.0.4-1) unstable; urgency=low

  * Changes for Kano OS 3.14.1:
  * Disabled hdmi_drive config option when going into Safe Mode
  * Make safe mode use the preferred mode of the screen
  * Fixed bug that causes no timeout on Wi-Fi failure

 -- Team Kano <dev@kano.me>  Tue, 16 Feb 2018 15:39:30 +0100

kano-settings (4.0.4-0) unstable; urgency=low

  * Changes for Kano OS 3.14.0:
  * Reduced execution time of kano-safeboot-mode reducing time to boot by ~3.5s
  * Migrated parental whitelist over to use the master branch
  * Added CKC speaker claim page to whitelist
  * Removed duplication of configuring /etc/hosts for Hack Minecraft
  * Added sentinel file on /boot partition to use custom config.txt settings
  * Write /boot/cpuid.txt
  * Ensure that the LightDM configuration exists on boot and regenerate if needed
  * Add kano zendesk to the parental control whitelist.

 -- Team Kano <dev@kano.me>  Tue, 5 Dec 2017 17:22:00 +0100

kano-settings (4.0.3-0) unstable; urgency=low

  * Changes for Kano OS 3.13.0:
  * Migrate Busybox dhcp client to a more modern implementation
  * Added Volume Boost option in Settings > Audio for CKCv1
  * Ensured volume levels at bootup for CKCv1 only
  * Included full screen model EDID in Settings > Display
  * Removed unused kano_screen_used config comment
  * Added edid-tools for dumping and flashing EDIDs
  * Added check for wireless passphrase lenght in GUI connect app

 -- Team Kano <dev@kano.me>  Thu, 12 Oct 2017 14:27:15 +0100

kano-settings (4.0.2-0) unstable; urgency=low

  * Changes for Kano OS 3.12.0:
  * Ensuring maximum volume level on CKC and other
  * Fixed resolution policy to account for preferred hdmi_group
  * Removed hdmi_drive=2 config option for old monitors
  * Added display configuration in config.txt for all our SKs
  * Added tests for display functions with an EDID dataset
  * Added CKC v1.1 EDID to tests and detection functions

 -- Team Kano <dev@kano.me>  Thu, 7 Sep 2017 18:11:09 +0100

kano-settings (4.0.1-0) unstable; urgency=low

  * Disable bootup sound init script, as alsa driver is disabled
  * Added the new Matrix screen saver selection into kano-settings
  * User preference in Advanced to enable or disable the SSH client
  * Mechanism to enable SSH client at boot with a /boot/ssh file
  * CPU id displayed in about section

 -- Team Kano <dev@kano.me>  Thu, 13 Jul 2017 17:41:09 +0100

kano-settings (4.0.0-0) unstable; urgency=low

  * Add Speed up boot times into Dashboard

 -- Team Kano <dev@kano.me>  Thu, 25 May 2017 16:38:00 +0100

kano-settings (3.9.1-1) unstable; urgency=low

  * Add API to detect the Kano Screen Kit

 -- Team Kano <dev@kano.me>  Thu, 30 Mar 2017 17:24:00 +0100

kano-settings (3.9.0-1) unstable; urgency=low

  * Add Minecraft local hostname

 -- Team Kano <dev@kano.me>  Tue 13 Dec 2016 13:08:00 +0000

kano-settings (3.8.0-1) unstable; urgency=low

  * Add kano-boot-check command to check for improper shutdown.

 -- Team Kano <dev@kano.me>  Thu, 13 Oct 2016 17:47:00 +0100

kano-settings (3.7.0-1) unstable; urgency=low

  * i18n setup and Spanish translations

 -- Team Kano <dev@kano.me>  Tue, 27 Sep 2016 11:30:00 +0100

kano-settings (3.2.0-1) unstable; urgency=low

  * Remove kano_settings.system.advanced.set_user_cookies API

 -- Team Kano <dev@kano.me>  Thu, 5 May 2016 11:00:00 +0100


kano-settings.py
#!/usr/bin/env python3

# http://python-gtk-3-tutorial.readthedocs.org/en/latest/layout.html

from gi.repository import Gtk, Gdk,GdkPixbuf, Pango
import set_intro
import set_email
import set_keyboard
import set_audio
import set_display
import set_wifi
import dialog_box
import config_file

win = None
MAX_STATE = 6
names = ["Email", "Keyboard", "Audio", "Wifi", "Display"]
custom_info = ["Email", "Keyboard-country-human", "Audio", "Wifi", "Display"]


# Window class
class MainWindow(Gtk.Window):
    grid = None
    box = None
    state = 0

    def __init__(self):
        global grid, box

        self.last_level_visited = 0

        TOP_BAR_HEIGHT = 50
        WINDOW_WIDTH = 600
        WINDOW_HEIGHT = 380

        # Create main window
        Gtk.Window.__init__(self, title="Kano-Settings")
        # Remove decoration
        self.set_decorated(False)

        self.set_size_request(WINDOW_WIDTH, WINDOW_HEIGHT)

        # Main container of the window
        self.grid = Gtk.Grid()

        # This is to give the correct colour of the top bar as Event Boxes are the only containers that we can colour
        self.top_bar_container = Gtk.EventBox()
        self.top_bar_container.set_size_request(WINDOW_WIDTH, TOP_BAR_HEIGHT)
        top_bar_container_style = self.top_bar_container.get_style_context()
        top_bar_container_style.add_class('top_bar_container')

        # Main title of he window bar.
        header = Gtk.Label("Kano Settings")
        header.props.halign = Gtk.Align.CENTER
        header.modify_font(Pango.FontDescription("Bariol 13"))
        header.get_style_context().add_class("header")

        self.top_bar = Gtk.Box()
        self.top_bar.set_homogeneous(False)

        pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size("media/Icons/systemsetup-icons.png", 192, 24)
        icons = []
        images = []

        for x in range(int(192/24)):
            subpixel = pixbuf.new_subpixbuf(24*x, 0, 24, 24).add_alpha(True,255,255,255)
            icons.append(subpixel)
            images.append(Gtk.Image())
            images[x].set_from_pixbuf(icons[x])

        # Prev Button
        #bPrev = Gtk.Button(label="Previous", halign=Gtk.Align.START)
        prev_button = Gtk.Button()
        prev_button.set_image(images[2])
        prev_button.set_size_request(TOP_BAR_HEIGHT, TOP_BAR_HEIGHT)
        prev_button.set_can_focus(False)

        # Next button
        next_button = Gtk.Button()
        next_button.set_image(images[3])
        next_button.set_size_request(TOP_BAR_HEIGHT, TOP_BAR_HEIGHT)
        next_button.set_can_focus(False)

        close_button = Gtk.Button()
        close_button.set_image(images[6])
        close_button.set_size_request(TOP_BAR_HEIGHT, TOP_BAR_HEIGHT)
        close_button.connect("clicked", Gtk.main_quit)
        close_button.set_can_focus(False)

        # Dynamic box

        #############################################################################
        # Move this into set_email etc (or separate file)
        # Needed for everything but intro screens.

        self.apply_changes_text = Gtk.Label("Get started")
        self.apply_changes = Gtk.Button()
        self.apply_changes.props.halign = Gtk.Align.CENTER
        self.apply_changes.set_size_request(70, 30)
        self.apply_changes_label = Gtk.Box()
        self.apply_changes_label.pack_start(self.apply_changes_text, False, False, 2)
        self.apply_changes_label.pack_start(images[0], False, False, 2)
        self.apply_changes.add(self.apply_changes_label)
        self.apply_changes_text.modify_font(Pango.FontDescription("Bariol bold 13"))
        self.apply_changes.connect('clicked', self.update_and_next)
        self.apply_changes.set_can_focus(False)

        next_button_style = self.apply_changes.get_style_context()
        next_button_style.add_class('apply_changes')

        self.changable_content = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=6)
        self.changable_content.props.halign = Gtk.Align.CENTER

        # Init
        if config_file.read_from_file('Completed') == "1":
            print("Settings has been completed before")
            self.default_intro = Gtk.Table(3, 2, True)
            self.default_intro_init()
            #Need new functions here
            next_button.connect("clicked", self.default_next)
            prev_button.connect("clicked", self.back_to_intro)
        else:
            print("Settings has NOT been completed before")
            set_intro.activate(self, self.changable_content, self.apply_changes)
            prev_button.connect("clicked", self.on_prev)
            next_button.connect("clicked", self.on_next)

        self.top_bar.add(prev_button)
        self.top_bar.pack_start(next_button, False, False, 0)
        self.top_bar.pack_start(header, True, True, 0)
        self.top_bar.pack_end(close_button, False, False, 0)
        self.top_bar_container.add(self.top_bar)
        self.grid.attach(self.top_bar_container, 0, 0, 1, 1)

        self.grid.attach(self.changable_content, 0, 2, 1, 1)
        self.grid.set_row_spacing(0)

        self.add(self.grid)


    def back_to_intro(self, arg2):
        # save last level?
        for i in self.changable_content.get_children():
            self.changable_content.remove(i)

        self.default_intro_text()

        self.changable_content.pack_start(self.default_intro, True, True, 0)


    def default_intro_init(self):
        global grid, box, state

        for i in self.changable_content.get_children():
            self.changable_content.remove(i)

        # Read from config file.
        # If value is 1:
        self.default_intro = Gtk.Table(3, 2, True)

        buttons = []
        self.labels = []

        #names at top of file
        for x in range(len(names)):
            label_box = Gtk.Box(orientation = Gtk.Orientation.VERTICAL, spacing=0)
            button_box = Gtk.Box(orientation = Gtk.Orientation.HORIZONTAL, spacing=0)

            label = Gtk.Label(names[x])
            label.get_style_context().add_class("intro_label")

            print(custom_info[x])
            info = config_file.read_from_file(custom_info[x])

            if info != None:
                # Replace some of the info displayed with whitespace so it fits
                if len(info) >= 12:
                    info = info[0:12] + '...'

            label2 = Gtk.Label(info)
            label2.get_style_context().add_class("custom_label")
            self.labels.append(label2)
            button = Gtk.Button()
            button.set_can_focus(False)
            img = Gtk.Image()
            img.set_from_file("media/Icons/Icon-" + names[x] + ".png")

            label_box.pack_start(label, True, True, 0)
            label_box.pack_start(label2, True, True, 0)
            label_box.set_size_request(90,20)
            #label.set_justify(Gtk.Justification.LEFT)
            #label2.set_justify(Gtk.Justification.LEFT)

            button_box.pack_start(img, True, True, 10)
            button_box.pack_start(label_box, True, True, 0)
            button_box.set_size_request(180, 80)
            button.add(button_box)
            buttons.append(button)
            button.state = x
            button.set_size_request(200, 100)
            button.connect("clicked", self.go_to_level)

        # Attach to table
        self.default_intro.attach(buttons[0], 0, 1, 0, 1, Gtk.AttachOptions.EXPAND, Gtk.AttachOptions.EXPAND, 5, 5)
        self.default_intro.attach(buttons[1], 0, 1, 1, 2, Gtk.AttachOptions.EXPAND, Gtk.AttachOptions.EXPAND, 5, 5)
        self.default_intro.attach(buttons[2], 0, 1, 2, 3, Gtk.AttachOptions.EXPAND, Gtk.AttachOptions.EXPAND, 5, 5)
        self.default_intro.attach(buttons[3], 1, 2, 0, 1, Gtk.AttachOptions.EXPAND, Gtk.AttachOptions.EXPAND, 5, 5)
        self.default_intro.attach(buttons[4], 1, 2, 1, 2, Gtk.AttachOptions.EXPAND, Gtk.AttachOptions.EXPAND, 5, 5)

        self.default_intro.set_size_request(450, 100)

        self.changable_content.pack_start(self.default_intro, True, True, 0)


    def default_intro_text(self):
        for x in range(len(custom_info)):
            config_file.read_from_file(custom_info[x])
            self.labels[x].set_text(str(config_file.read_from_file(custom_info[x])))


    def update_and_next(self, widget):
        returnValue = self.state_to_widget(self.state).apply_changes(widget)
        if returnValue == -1:
            return

        self.on_next(widget)


    def default_next(self, widget):
        if self.last_level_visited == 0:
            return

        for i in self.changable_content.get_children():
            self.changable_content.remove(i)

        
        self.state_to_widget(self.last_level_visited).activate(self, self.changable_content, self.apply_changes) 
        self.last_level_visited = self.state   
        win.show_all()

    def on_next(self, widget):
        global grid, box, state

        # Update current state
        self.state = (self.state + 1)
        # If we've clicked on the finished
        if self.state == MAX_STATE:
            # Write to config file to say we've completed the level.
            config_file.replace_setting("Completed", "1")
            # Finished, so close window
            close_window()
            return

        # Remove element in the dynamic box
        for i in self.changable_content.get_children():
            self.changable_content.remove(i)

        # Call next state
        self.state_to_widget(self.state).activate(self, self.changable_content, self.apply_changes)

        # Change label on Apply Changes button
        if self.state == MAX_STATE -1:
            self.apply_changes_text.set_text("Finish!")
        elif self.state > 0:
            self.apply_changes_text.set_text("Next")
        else:
            self.apply_changes_text.set_text("Get started")
        # Refresh window
        win.show_all()

    def on_prev(self, widget):
         # Update current state
        self.state = (self.state - 1)
        # Check we're in a valid state
        if self.state == -1:
            self.state = 0
            return;

        # Remove element in the dynamic box
        for i in self.changable_content.get_children():
            self.changable_content.remove(i)
       
        # Call next state
        self.state_to_widget(self.state).activate(self, self.changable_content, self.apply_changes)

        # Change label on Apply Changes button
        if self.state == MAX_STATE -1:
            self.apply_changes_text.set_text("Finish!")
        elif self.state > 0:
            self.apply_changes_text.set_text("Next")
        else:
            self.apply_changes_text.set_text("Get started")
        # Refresh window
        win.show_all()

    def go_to_level(self, widget):
        # Remove element in the dynamic box
        for i in self.changable_content.get_children():
            self.changable_content.remove(i)
        # Update current state
        self.state = widget.state + 1
        # Record this level so we can go back to it
        self.last_level_visited = self.state
        # Call next state
        self.state_to_widget(self.state).activate(self, self.changable_content, self.apply_changes)
        # Refresh window
        win.show_all()


    def state_to_widget(self, x):
        return {
            0: set_intro,
            1: set_email,
            2: set_keyboard,
            3: set_audio,
            4: set_wifi,
            5: set_display,
        }[x]


def main():
    global win
    # Create style sheet
    cssProvider = Gtk.CssProvider()
    cssProvider.load_from_path('style.css')
    screen = Gdk.Screen.get_default()
    styleContext = Gtk.StyleContext()
    styleContext.add_provider_for_screen(screen, cssProvider,
                                     Gtk.STYLE_PROVIDER_PRIORITY_USER)

    # Create windown
    win = MainWindow()
    # Link delete event
    win.connect("delete-event", close_window)
    # Display window
    win.show_all()

    # start the GTK+ processing loop
    Gtk.main()

def close_window(event="delete-event", button=win):

    if set_audio.reboot or set_display.reboot:
        #Bring in message dialog box
        dialog = Gtk.MessageDialog(self, 0, Gtk.MessageType.INFO,
            Gtk.ButtonsType.OK, "This is an INFO MessageDialog")
        dialog.format_secondary_text(
            "And this is the secondary text that explains things.")
        dialog.run()
        print("INFO dialog closed")

        if response == Gtk.ResponseType.OK:
            dialog.destroy() 
            Gtk.main_quit()
            return
        else:
            dialog.destroy()

    else:
        Gtk.main_quit()


if __name__ == "__main__":
    # Print system python version for debugging purposes only
    print(str(Gtk.get_major_version()) + "." + str(Gtk.get_minor_version()) + "." + str(Gtk.get_micro_version()))

    main()
